### ISSUE TITLE
[REQ] Automatic virtual gears

### ISSUE THREAD

**Issue by smulrine** [cagnulein/qdomyos-zwift#2236]
I usually change virtual gears in Zwift so that gradient - virtual gear = 3, so if I'm on a 9% incline, I change my virtual gear to -6, or -5 if I want to work a bit harder. When my boy is cycling I change his gears so that gradient - virtual gear = 0. It occurs to me that this could be automatic, with a toggle in the QZ settings to enable automatic gears, and pressing + or - on Zwift Click/Play would instead change N where N = gradient - virtual gear. I imagine something like ERG mode except with constant resistance instead of power, and it would certainly save a lot of button-pressing when going over rolling terrain.

**Comment by cagnulein** [comment]
you can use profiles + save gears on startup in the advanced settings for
this

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 17 mar 2024 alle 17:47 smulrine ***@***.***> ha
scritto:

**Comment by smulrine** [comment]
I know about profiles for different users, but the main idea is that QZ could change virtual gears automatically based on the gradient. If I have my target effective gear set to 3, then when I am on a 5% gradient, QZ will automatically apply a virtual gear of -2, and when I am on the flat, QZ will automatically apply a virtual gear of 3, without me having to press any Zwift Click buttons myself.

**Comment by cagnulein** [comment]
yes that's the qz default behavior. which bike do you have?

**Comment by smulrine** [comment]
I'm using a Tacx Neo. Virtual gears currently behave as I would expect; I have to change gear myself using Zwift Click instead of using physical gears on the bike, e.g when going uphill, and it's a decision I make because of the gradient shown on the screen (and because resistance has increased). What I am suggesting is that QZ could change gear for me as it already knows what the gradient is. Instead of transmitting gradient minus virtual gear to the trainer, which I believe is what currently happens, it transmits my target virtual gear to the trainer, which in my example would be 3. (OK, so it's not actually changing gear, it's sending the same gradient all the time!) Then the buttons could allow me to adjust my target virtual gear depending on what resistance I want.

So, in other words, "target virtual gear" is the gradient I want QZ to send, regardless of the terrain.

Now that I think about it, this would be exactly like setting Trainer Difficulty to zero in Zwift, but then the Elite Rizer and Wahoo Climb wouldn't work.

I think this would actually be a pretty easy coding tweak so I might give it a try myself to see if it works as I expect.

**Comment by cagnulein** [comment]
i don't understand, it's what qz is already doing so i guess it's a qz bug?

i mean qz simply does this

getting the gradient from zwift and send to the trainer  + gears all the
time.

So it's already doing what you are asking.

so could you please recreate the issue so i could check that i don't
missing something?

https://github.com/cagnulein/qdomyos-zwift/wiki/How-do-i-get-the-debug-log-in-case-something-doesn't-work%3F

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 17 mar 2024 alle 22:12 smulrine ***@***.***> ha
scritto:

**Comment by smulrine** [comment]
Definitely not a bug, I probably just didn't explain it very well, because I didn't quite understand myself how it would work. What I want is for QZ to send the same gradient to the trainer all the time, no matter what the actual gradient is - while also sending the real gradient to the Elite Rizer. It looks like automatic virtual gear changing, but in fact it's doing no calculations at all.

Maybe there are some settings which would already do this. I will take a closer look tomorrow.

**Comment by cagnulein** [comment]
hah so you simply don't want auto resistance? or maybe you would like to
use erg mode on a free ride?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 17 mar 2024 alle 22:41 smulrine ***@***.***> ha
scritto:

**Comment by smulrine** [comment]
Yes, I guess I don't really want auto-resistance since I am only changing gear to keep the resistance the same, more or less. It's not exactly ERG mode since that keeps the power the same, and if I pedal harder the resistance decreases. With no auto-resistance, if I pedal harder, the power increases.

**Comment by cagnulein** [comment]
If you put the zwift resitance gain to 0 you're obtaining what you need

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 17 mar 2024 alle 23:29 smulrine ***@***.***> ha
scritto:

**Comment by smulrine** [comment]
In fact zwift_inclination_gain = 0.05 and elite_rizer_gain = 20 does exactly what I want. Now I can also set gears_gain to 0.5 as I should never need to make a big gear change again. Thanks for your help!

**Comment by cagnulein** [comment]
perfect!
